R/batch_container.R

```diff
@@ -328,23 +328,67 @@ BatchContainer <- R6::R6Class("BatchContainer",
 
     #' @description
     #' Score current sample assignment,
-    #' @return Returns a vector of all scoring functions values.
-    score = function() {
-      assertthat::assert_that(!is.null(private$scoring_funcs),
-        msg = "Scoring function needs to be assigned"
+    #' @param scoring a function or a names list of scoring functions.
+    #' Each function should return a numeric vector.
+    #' @return Returns a named vector of all scoring functions values.
+    score = function(scoring) {
+      assertthat::assert_that(
+        !missing(scoring),
+        !is.null(scoring),
+        msg = "Scoring function needs to be provided"
       )
-      assertthat::assert_that(is.list(private$scoring_funcs),
-        length(private$scoring_funcs) >= 1,
-        msg = "Scroring function should be a non-empty list"
+      if (is.function(scoring)) {
+        scoring <- list(scoring)
+      } else {
+        assertthat::assert_that(is.list(scoring), length(scoring) >= 1)
+        assertthat::assert_that(
+          all(purrr::map_lgl(scoring, is.function)),
+          msg = "All elements of scoring should be functions"
+        )
+      }
+      if (is.null(names(scoring))) {
+        names(scoring) <- stringr::str_c("score_", seq_along(scoring))
+      }
+      assertthat::assert_that(
+        !any(names(scoring) == ""),
+        msg = "scoring cannot be a partially named list"
       )
-      assertthat::assert_that(!is.null(private$samples_table),
+      assertthat::assert_that(is.list(scoring),
+        length(scoring) >= 1,
+        msg = "Scoring function should be a non-empty list"
+      )
+      assertthat::assert_that(!is.null(names(scoring)),
+        msg = "scoring should be a named list"
+      )
+      assertthat::assert_that(self$has_samples,
         msg = "No samples in the batch container, cannot compute score"
       )
-
-      res <- purrr::map_dbl(private$scoring_funcs, ~ .x(self))
-      assertthat::assert_that(length(res) == length(private$scoring_funcs))
-
-      assertthat::assert_that(is.numeric(res), msg = "Scoring function should return a number")
+      res <- purrr::imap(
+        scoring,
+        \(f, i) {
+          v <- f(self)
+          assertthat::assert_that(
+            is.numeric(v),
+            length(v) >= 1,
+            msg = "scoring function should return a numeric vector of positive length"
+          )
+          if (length(v) > 1) {
+            if (is.null(names(v))) {
+              names(v) <- seq_along(v)
+            }
+            names(v) <- stringr::str_c(i, names(v))
+          } else {
+            names(v) <- i
+          }
+          v
+        }
+      ) |>
+        purrr::flatten_dbl()
+      assertthat::assert_that(length(res) >= length(scoring))
+      assertthat::assert_that(
+        !any(names(res) == "step"),
+        msg = "score name cannot be 'step'"
+      )
```

Comment:
minor nitpicking: shouldn't this be checked before the scoring is run? (for performance reasons)

Reply:
I think there's no way of knowing, because score names could come from the scoring function; it could return a named vector. you wouldn't know before you call it.

Reply:
ah, true, let's keep it like this